File: android/mekuri/src/main/kotlin/studio/anicanon/mekuri/MekuriFoldGeometry.kt
package studio.anicanon.mekuri

import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/**
 * Pure fold arithmetic in page space. Progress runs 0 (flat) to 1 (turned); the
 * fold axis travels from the trailing edge to the leading edge. Fold distance is
 * measured from the held end, `y == pageHeight`, toward the free corner at
 * `y == 0`.
 *
 * These formulas are mirrored in the fold shader and are held in numeric parity
 * with the iOS geometry.
 */
internal data class MekuriFoldGeometry(
    val pageWidth: Float,
    val configuration: MekuriConfiguration = MekuriConfiguration.Default,
) {
    /** Cylinder radius at the held end. */
    val heldRadius: Float
        get() = pageWidth * configuration.cylinderRadiusRatio

    /** Radius growth per unit of fold distance. */
    val radiusSlope: Float
        get() = configuration.cylinderRadiusRatio * configuration.radiusOpening

    /**
     * Horizontal position of the fold axis at the page's vertical centre with no
     * bow. Out-of-range progress clamps to 0..1.
     */
    fun foldAxisOffset(progress: Float): Float = pageWidth * (1 - clamped(progress))

    /** Horizontal position of the crease at [y], including shear and bow. */
    fun foldAxisOffset(progress: Float, y: Float, pageHeight: Float): Float {
        val held = 1 - y / pageHeight
        val shear = configuration.cornerShear * (y - pageHeight / 2)
        return foldAxisOffset(progress) + shear - bowLead(progress) * held * held
    }

    /** Distance the free corner runs ahead of the straight crease. */
    fun bowLead(progress: Float): Float {
        val clamped = clamped(progress)
        return configuration.creaseBow * pageWidth * clamped * (1 - clamped)
    }

    /** Cylinder radius at [foldDistance] along the fold from the held end. */
    fun radius(foldDistance: Float): Float = heldRadius + radiusSlope * foldDistance

    fun isFlat(progress: Float): Boolean = progress <= 0f

    /**
     * Scale on the cylinder radius and corner shear: 1 until the landing
     * begins, [LandingFloor] at progress 1.
     */
    fun landingRadiusScale(progress: Float): Float {
        val remaining = 1 - clamped(progress)
        return max(min(remaining / LandingFraction, 1f), LandingFloor)
    }

    /** Full-height band centred on the fold axis. */
    fun creaseShadowRect(progress: Float, pageSize: Size): Rect {
        val width = pageWidth * configuration.creaseShadowWidthRatio
        val axis = foldAxisOffset(progress)
        return Rect(
            left = axis - width / 2,
            top = 0f,
            right = axis + width / 2,
            bottom = pageSize.height,
        )
    }

    /**
     * Horizontal position of the sheet's free edge at row [y] of the shader's
     * frame, as the shader draws it: on the roll while the flap is shorter than
     * half a turn, lying flat past the crest beyond. Radius and shear land with
     * the turn.
     */
    fun freeEdge(progress: Float, y: Float, pageHeight: Float): Float {
        val landing = landingRadiusScale(progress)
        val held = 1 - y / pageHeight
        val shear = configuration.cornerShear * landing * (y - pageHeight / 2)
        val axis = foldAxisOffset(progress) + shear - bowLead(progress) * held * held
        val flap = pageWidth - axis
        val radius = radius(pageHeight - y) * landing
        val halfTurn = PI.toFloat() * radius
        return if (flap <= halfTurn) axis + radius * sin(flap / radius) else axis - (flap - halfTurn)
    }

    /**
     * Progress at which [freeEdge] at row [y] reaches [edge]. The edge only moves
     * toward the spine as progress grows.
     */
    fun progressForFreeEdge(edge: Float, y: Float, pageHeight: Float): Float {
        var low = 0f
        var high = 1f
        repeat(EdgeSearchSteps) {
            val mid = (low + high) / 2
            if (freeEdge(mid, y, pageHeight) > edge) low = mid else high = mid
        }
        return (low + high) / 2
    }

    private fun clamped(progress: Float): Float = min(max(progress, 0f), 1f)

    companion object {
        /** Bisection steps when inverting [freeEdge]; fixed so every platform lands on the same value. */
        const val EdgeSearchSteps = 32

        /** Share of the turn over which the roll flattens. */
        const val LandingFraction = 0.3f

        /** Smallest radius scale; the contact shadow divides by the radius. */
        const val LandingFloor = 0.01f
    }
}

File: ios/Sources/Mekuri/MekuriFoldGeometry.swift
import CoreGraphics
import Foundation

/// Pure fold arithmetic in page space. Progress runs 0 (flat) to 1 (turned);
/// the fold axis travels from the trailing edge to the leading edge. Fold
/// distance is measured from the held end, `y == pageHeight`, toward the free
/// corner at `y == 0`. These formulas are mirrored in `mekuriFold`.
struct MekuriFoldGeometry: Equatable, Sendable {
    let pageWidth: CGFloat
    let configuration: MekuriConfiguration

    init(pageWidth: CGFloat, configuration: MekuriConfiguration) {
        self.pageWidth = pageWidth
        self.configuration = configuration
    }

    /// Horizontal position of the fold axis at the page's vertical centre
    /// with no bow. Out-of-range progress clamps to 0...1.
    func foldAxisOffset(progress: CGFloat) -> CGFloat {
        self.pageWidth * (1 - self.clamped(progress))
    }

    /// Horizontal position of the crease at `y`, including shear and bow.
    func foldAxisOffset(progress: CGFloat, y: CGFloat, pageHeight: CGFloat) -> CGFloat {
        let held = 1 - y / pageHeight
        let shear = self.configuration.cornerShear * (y - pageHeight / 2)
        return self.foldAxisOffset(progress: progress) + shear - self.bowLead(progress: progress) * held * held
    }

    /// Distance the free corner runs ahead of the straight crease.
    func bowLead(progress: CGFloat) -> CGFloat {
        let progress = self.clamped(progress)
        return self.configuration.creaseBow * self.pageWidth * progress * (1 - progress)
    }

    /// Cylinder radius at `distance` points along the fold from the held end.
    func radius(atFoldDistance distance: CGFloat) -> CGFloat {
        self.heldRadius + self.radiusSlope * distance
    }

    /// Cylinder radius at the held end.
    var heldRadius: CGFloat {
        self.pageWidth * self.configuration.cylinderRadiusRatio
    }

    /// Radius growth per point of fold distance.
    var radiusSlope: CGFloat {
        self.configuration.cylinderRadiusRatio * self.configuration.radiusOpening
    }

    func isFlat(progress: CGFloat) -> Bool {
        progress <= 0
    }

    /// Share of the turn over which the roll flattens.
    static let landingFraction: CGFloat = 0.3

    /// Smallest radius scale; the contact shadow divides by the radius.
    static let landingFloor: CGFloat = 0.01

    /// Scale on the cylinder radius and corner shear: 1
    /// until the landing begins, `landingFloor` at progress 1.
    func landingRadiusScale(progress: CGFloat) -> CGFloat {
        let remaining = 1 - self.clamped(progress)
        return max(min(remaining / Self.landingFraction, 1), Self.landingFloor)
    }

    /// Full-height band centred on the fold axis.
    func creaseShadowRect(progress: CGFloat, pageSize: CGSize) -> CGRect {
        let width = self.pageWidth * self.configuration.creaseShadowWidthRatio
        let axis = self.foldAxisOffset(progress: progress)
        return CGRect(x: axis - width / 2, y: 0, width: width, height: pageSize.height)
    }

    /// Horizontal position of the sheet's free edge at row `y` of the
    /// shader's frame, as the shader draws it: on the roll while the flap is
    /// shorter than half a turn, lying flat past the crest beyond. Radius and
    /// shear land with the turn.
    func freeEdge(progress: CGFloat, y: CGFloat, pageHeight: CGFloat) -> CGFloat {
        let landing = self.landingRadiusScale(progress: progress)
        let held = 1 - y / pageHeight
        let shear = self.configuration.cornerShear * landing * (y - pageHeight / 2)
        let axis = self.foldAxisOffset(progress: progress) + shear - self.bowLead(progress: progress) * held * held
        let flap = self.pageWidth - axis
        let radius = self.radius(atFoldDistance: pageHeight - y) * landing
        let halfTurn = CGFloat.pi * radius
        return flap <= halfTurn ? axis + radius * sin(flap / radius) : axis - (flap - halfTurn)
    }

    private func clamped(_ progress: CGFloat) -> CGFloat {
        min(max(progress, 0), 1)
    }
}
